File: RealtimeFeed/Sources/RealtimeFeed/ReconnectPolicy.swift
import Foundation

/// Pure exponential backoff with jitter, kept separate from I/O so it's trivially unit-testable.
public struct ReconnectPolicy: Sendable {
    public let baseDelay: TimeInterval
    public let maxDelay: TimeInterval
    public let jitterFraction: Double

    public init(baseDelay: TimeInterval = 1, maxDelay: TimeInterval = 30, jitterFraction: Double = 0.3) {
        self.baseDelay = baseDelay
        self.maxDelay = maxDelay
        self.jitterFraction = jitterFraction
    }

    /// attempt starts at 1 for the first reconnect try.
    public func delay(forAttempt attempt: Int, randomSource: () -> Double = { Double.random(in: 0...1) }) -> TimeInterval {
        let exponential = baseDelay * pow(2, Double(max(0, attempt - 1)))
        let capped = min(maxDelay, exponential)
        let jitter = capped * jitterFraction * randomSource()
        return capped + jitter
    }
}

File: shared/src/commonMain/kotlin/dev/web3demo/realtimefeed/ReconnectPolicy.kt
package dev.web3demo.realtimefeed

import kotlin.math.min
import kotlin.math.pow

/**
 * Pure exponential backoff with jitter, kept separate from I/O so it's trivially unit-testable.
 * Mirrors the Swift `ReconnectPolicy` in RealtimeFeed/ 1:1 so both platforms reconnect the same way.
 */
class ReconnectPolicy(
    private val baseDelayMillis: Long = 1_000,
    private val maxDelayMillis: Long = 30_000,
    private val jitterFraction: Double = 0.3,
) {
    /** attempt starts at 1 for the first reconnect try. */
    fun delayMillis(attempt: Int, randomSource: () -> Double = { kotlin.random.Random.nextDouble() }): Long {
        val exponential = baseDelayMillis * 2.0.pow((attempt - 1).coerceAtLeast(0))
        val capped = min(maxDelayMillis.toDouble(), exponential)
        val jitter = capped * jitterFraction * randomSource()
        return (capped + jitter).toLong()
    }
}
